File: Projects/leetcode/Jump.java
package leetcode;

class MaxResult {
    public static int maxResult(int[] nums, int k) {
        final int n = nums.length;
        final int[] cache = new int[n];
        cache[0] = nums[0];
        for (int i = 1; i < n; i++) {
            cache[i] = Integer.MIN_VALUE;
        }
        for (int i = 1; i < n; i++) {
            final int num = nums[i];
            for (int j = 1; j <= Math.min(k, i); j++) {
                cache[i] = Math.max(cache[i], num + cache[i - j]);
            }
        }
        return cache[n - 1];
    }
}

File: Projects/leetcode/jump.kt
package leetcode

fun jump(nums: IntArray): Int {
    if (nums.size == 1) return 0
    var jump = 0
    var curr = 0
    var next = 0
    for ((idx, num) in nums.withIndex()) {
        if (idx + num >= nums.lastIndex) {
            return jump + 1
        }
        if (idx + num > next) {
            next = idx + num
        }
        if (idx == curr) {
            curr = next
            jump += 1
        }
    }
    return jump
}
